File: flutter/lib/src/erghi_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:web_socket_channel/web_socket_channel.dart';
import 'config/erghi_config.dart';
import 'resources/resources.dart';
import 'models/models.dart';
import 'exceptions/exceptions.dart';

/// ASP.NET Core SignalR's JSON Hub Protocol terminates every frame with this character.
const String _recordSeparator = '\x1e';

/// Real hub broadcast names (Erghi.Conversation/Api/Hubs/ChatHub.cs /
/// ConversationRealtimeNotifier.cs) this client listens for.
const String _messageReceivedTarget = 'MessageReceived';

/// Main Erghi SDK Client
class ErghiClient {
  final ErghiConfig config;
  late final http.Client _httpClient;
  late final AuthResource auth;
  late final ChatResource chat;

  WebSocketChannel? _wsChannel;
  StreamSubscription? _wsSubscription;
  StreamController<Message>? _messageController;
  Timer? _keepaliveTimer;
  bool _isConnected = false;
  final Map<String, Completer<dynamic>> _pendingInvocations = {};
  String? visitorId;

  ErghiClient({required this.config, http.Client? httpClient}) {
    final innerClient = httpClient ?? http.Client();
    final m2mClient = M2MHttpClient(innerClient, config);
    _httpClient = m2mClient;
    auth = AuthResource(config: config, client: _httpClient);
    m2mClient.auth = auth;
    chat = ChatResource(
      config: config,
      client: _httpClient,
      auth: auth,
      getVisitorId: () => visitorId,
    );
  }

  /// Check if the real-time hub is connected
  bool get isConnected => _isConnected;

  /// Connect to the real-time hub for real-time updates.
  ///
  /// Was a raw `web_socket_channel` connection speaking a hand-rolled {type, conversationId}
  /// envelope directly at /hubs/chat -- that never completes the SignalR negotiate/handshake a
  /// real ASP.NET Core SignalR hub requires (a POST to /negotiate for a connection token, then
  /// a JSON handshake record before any invocation frames are accepted), so it could not
  /// actually exchange messages with the real backend. This implements that protocol directly
  /// on top of the same `web_socket_channel` package, since no official/maintained SignalR
  /// client exists for Dart/Flutter. Also fixes a standing bug where the auth headers this
  /// method built were never actually attached to the connection -- auth now travels the way
  /// the real hub actually expects it (an `access_token` query parameter, the same mechanism
  /// browsers use since they can't set custom headers on a WebSocket upgrade either).
  Future<void> connectWebSocket() async {
    if (_isConnected) return;

    try {
      final connectionToken = await _negotiate();
      final accessToken = config.apiKey ?? auth.accessToken ?? '';
      final wsUri = Uri.parse('${config.websocketUrl}/hubs/chat').replace(
        queryParameters: {
          'id': connectionToken,
          'access_token': accessToken,
        },
      );

      _wsChannel = WebSocketChannel.connect(wsUri);
      await _wsChannel!.ready;

      _messageController = StreamController<Message>.broadcast();

      // WebSocketChannel.stream is single-subscription, so the handshake response and every
      // later frame have to flow through the one listener below -- a separate temporary
      // .listen() call just for the handshake (as a first draft of this method did) throws
      // "Stream has already been listened to" the moment the real frame listener attaches.
      final handshakeCompleter = Completer<void>();
      var handshakeDone = false;

      _wsSubscription = _wsChannel!.stream.listen(
        (data) {
          if (!handshakeDone) {
            handshakeDone = true;
            final text = data is List<int> ? utf8.decode(data) : data as String;
            final record = text.split(_recordSeparator).first;
            final response = record.isEmpty ? <String, dynamic>{} : jsonDecode(record) as Map<String, dynamic>;
            if (response['error'] != null) {
              handshakeCompleter.completeError(WebSocketException('Hub handshake failed: ${response['error']}'));
            } else {
              _isConnected = true;
              handshakeCompleter.complete();
            }
            return;
          }
          _handleFrame(data);
        },
        onError: (error) {
          if (config.debug) {
            print('Hub error: $error');
          }
          _isConnected = false;
          if (!handshakeCompleter.isCompleted) handshakeCompleter.completeError(error);
        },
        onDone: () {
          _isConnected = false;
          if (config.debug) {
            print('Hub connection closed');
          }
        },
      );

      _wsChannel!.sink.add(jsonEncode({'protocol': 'json', 'version': 1}) + _recordSeparator);
      await handshakeCompleter.future;

      _keepaliveTimer = Timer.periodic(const Duration(seconds: 15), (_) {
        if (_isConnected) {
          _wsChannel?.sink.add(jsonEncode({'type': 6}) + _recordSeparator);
        }
      });
    } catch (e) {
      throw WebSocketException('Failed to connect to hub: $e');
    }
  }

  /// POST /hubs/chat/negotiate to obtain the connection token the WebSocket URL needs.
  Future<String> _negotiate() async {
    final uri = Uri.parse('${config.apiUrl}/hubs/chat/negotiate')
        .replace(queryParameters: {'negotiateVersion': '1'});
    final headers = <String, String>{};
    if (config.apiKey != null) {
      headers['X-API-Key'] = config.apiKey!;
    } else if (auth.accessToken != null) {
      headers['Authorization'] = 'Bearer ${auth.accessToken}';
    }

    final response = await _httpClient.post(uri, headers: headers);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw WebSocketException('Hub negotiate failed: ${response.statusCode}');
    }

    final data = jsonDecode(response.body) as Map<String, dynamic>;
    if (data['url'] != null) {
      // Azure SignalR-style redirect to a different endpoint -- not used by this platform's
      // self-hosted deployment, but fail clearly instead of silently connecting elsewhere.
      throw WebSocketException('Hub negotiate returned a redirect target; not supported');
    }

    final token = data['connectionToken'] ?? data['connectionId'];
    if (token == null) {
      throw WebSocketException('Hub negotiate response had no connection token');
    }
    return token.toString();
  }

  void _handleFrame(dynamic data) {
    final text = data is List<int> ? utf8.decode(data) : data as String;
    for (final record in text.split(_recordSeparator)) {
      if (record.isEmpty) continue;
      try {
        _handleRecord(jsonDecode(record) as Map<String, dynamic>);
      } catch (e) {
        if (config.debug) {
          print('Failed to parse hub message: $e');
        }
      }
    }
  }

  void _handleRecord(Map<String, dynamic> message) {
    final type = message['type'];

    if (type == 1) {
      // Invocation from the server -- a hub broadcast (MessageReceived, UserTyping, ...).
      final target = message['target'];
      final arguments = (message['arguments'] as List?) ?? const [];
      if (target == _messageReceivedTarget && arguments.isNotEmpty) {
        try {
          _messageController?.add(Message.fromJson(arguments[0] as Map<String, dynamic>));
        } catch (e) {
          if (config.debug) {
            print('Failed to parse MessageReceived payload: $e');
          }
        }
      }
    } else if (type == 3) {
      // Completion -- the response to one of our own invoke() calls.
      final invocationId = message['invocationId'];
      final completer = invocationId != null ? _pendingInvocations.remove(invocationId) : null;
      if (completer != null && !completer.isCompleted) {
        if (message['error'] != null) {
          completer.completeError(HubException(message['error'].toString()));
        } else {
          completer.complete(message['result']);
        }
      }
    } else if (type == 7) {
      // Server-initiated close.
      if (config.debug) {
        print('Hub sent close: ${message['error']}');
      }
      _isConnected = false;
    }
    // Type 6 (ping) needs no application-level response beyond our own keepalive timer.
    // Types 2/4/5 (streaming) are not used by ChatHub/VisitorChatHub.
  }

  /// Invoke a real hub method (e.g. "SendTyping", "JoinConversation", "MarkAsRead").
  ///
  /// When [waitForResult] is true (the default), waits for the server's Completion message and
  /// throws [HubException] if the server rejected the call -- notably including
  /// ConversationOwnershipHubFilter denying access to a conversation outside the caller's own
  /// workspace, which the previous transport could never have surfaced since it never spoke to
  /// a real hub at all.
  Future<dynamic> invoke(String method, List<dynamic> arguments, {bool waitForResult = true}) async {
    if (!_isConnected || _wsChannel == null) {
      throw WebSocketException('Hub is not connected');
    }

    final message = <String, dynamic>{'type': 1, 'target': method, 'arguments': arguments};

    Completer<dynamic>? completer;
    String? invocationId;
    if (waitForResult) {
      invocationId = '${DateTime.now().microsecondsSinceEpoch}-$method';
      message['invocationId'] = invocationId;
      completer = Completer<dynamic>();
      _pendingInvocations[invocationId] = completer;
    }

    _wsChannel!.sink.add(jsonEncode(message) + _recordSeparator);

    if (completer != null) {
      try {
        return await completer.future.timeout(config.timeout);
      } finally {
        if (invocationId != null) {
          _pendingInvocations.remove(invocationId);
        }
      }
    }
    return null;
  }

  /// Disconnect from the hub.
  Future<void> disconnectWebSocket() async {
    _keepaliveTimer?.cancel();
    _keepaliveTimer = null;
    await _wsSubscription?.cancel();
    _wsSubscription = null;
    for (final completer in _pendingInvocations.values) {
      if (!completer.isCompleted) completer.completeError(WebSocketException('Hub disconnected'));
    }
    _pendingInvocations.clear();
    await _wsChannel?.sink.close();
    await _messageController?.close();
    _wsChannel = null;
    _messageController = null;
    _isConnected = false;
  }

  /// Stream of real-time messages (MessageReceived broadcasts).
  Stream<Message>? get messageStream => _messageController?.stream;

  /// Send typing indicator. Fire-and-forget, matching this method's previous behavior.
  void sendTyping(String conversationId) {
    if (!_isConnected) return;
    invoke('SendTyping', [conversationId], waitForResult: false).catchError((e) {
      if (config.debug) {
        print('Failed to send typing indicator: $e');
      }
    });
  }

  /// Join a conversation's real-time group -- required before UserTyping/MessageRead events
  /// for that conversation are delivered to this connection. Throws [HubException] if the
  /// conversation isn't in the caller's own workspace.
  Future<void> joinConversation(String conversationId) async {
    if (!_isConnected) return;
    await invoke('JoinConversation', [conversationId]);
  }

  /// Leave a conversation's real-time group.
  Future<void> leaveConversation(String conversationId) async {
    if (!_isConnected) return;
    await invoke('LeaveConversation', [conversationId], waitForResult: false);
  }

  /// Authenticate a visitor using a signed JWT from the customer's backend.
  Future<String> authenticateVisitor(String widgetId, String jwtToken) async {
    try {
      final url = Uri.parse('${config.apiUrl}/api/conversations/identity');
      final response = await _httpClient.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'widgetId': widgetId,
          'jwtToken': jwtToken,
        }),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final data = jsonDecode(response.body);
        visitorId = data['visitorId']?.toString() ?? data['VisitorId']?.toString();
        if (visitorId != null) {
          return visitorId!;
        }
        throw AuthenticationException('Visitor ID not found in identity response.');
      } else {
        throw AuthenticationException('Failed to authenticate visitor: ${response.body}');
      }
    } catch (e) {
      if (e is AuthenticationException) rethrow;
      throw AuthenticationException('Failed to authenticate visitor: $e');
    }
  }

  /// Dispose the client and clean up resources
  void dispose() {
    disconnectWebSocket();
    _httpClient.close();
  }
}

class M2MHttpClient extends http.BaseClient {
  final http.Client _inner;
  final ErghiConfig config;
  AuthResource? auth;

  M2MHttpClient(this._inner, this.config);

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    if (request.url.path.endsWith('/api/v1/auth/token')) {
      return _inner.send(request);
    }

    if (config.clientId != null && config.clientSecret != null && auth?.accessToken == null) {
      try {
        await auth?.authenticate();
      } catch (e) {
        // Auto-auth failed
      }
    }

    if (config.workspaceId != null) {
      request.headers['X-Workspace-Id'] = config.workspaceId!;
    }

    if (config.accountId != null) {
      request.headers['X-Account-Id'] = config.accountId!;
    }

    return _inner.send(request);
  }
}

File: swift/Sources/ErghiSDK/ErghiClient.swift
import Foundation
import Alamofire
import Starscream
import Combine

/// Main Erghi SDK Client
public class ErghiClient {
    /// ASP.NET Core SignalR's JSON Hub Protocol terminates every frame with this character.
    private static let recordSeparator = "\u{1e}"

    public let config: ErghiConfig
    public let auth: AuthResource
    public private(set) lazy var chat: ChatResource = ChatResource(
        config: config, session: session, auth: auth
    ) { [weak self] in
        self?.visitorId
    }

    private let session: Session
    private var webSocket: WebSocket?
    private let messageSubject = PassthroughSubject<Message, Never>()

    private var handshakeCompleted = false
    private var handshakeContinuation: CheckedContinuation<Void, Error>?
    private var pendingInvocations: [String: CheckedContinuation<Any?, Error>] = [:]
    private var keepaliveTask: Task<Void, Never>?

    public private(set) var visitorId: String?

    /// Stream of real-time messages
    public var messagePublisher: AnyPublisher<Message, Never> {
        messageSubject.eraseToAnyPublisher()
    }

    /// Check if the real-time hub is connected
    public private(set) var isConnected = false

    public init(config: ErghiConfig) {
        self.config = config

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = config.timeout

        let interceptor = M2MRequestInterceptor(config: config)
        self.session = Session(configuration: configuration, interceptor: interceptor)

        self.auth = AuthResource(config: config, session: session)

        interceptor.auth = auth
    }

    /// Connect to the real-time hub for real-time updates.
    ///
    /// Was a raw Starscream `WebSocket` speaking a hand-rolled {type, conversationId} envelope
    /// directly at /hubs/chat -- that never completes the SignalR negotiate/handshake a real
    /// ASP.NET Core SignalR hub requires (a POST to /negotiate for a connection token, then a
    /// JSON handshake record before any invocation frames are accepted), so it could not
    /// actually exchange messages with the real backend. This implements that protocol
    /// directly on top of the same Starscream connection, since no official/maintained
    /// SignalR client exists for Swift. Also fixes a standing bug where the auth headers this
    /// method built were never actually attached to the connection (Starscream/WebSocket
    /// upgrade requests don't reliably carry custom headers across platforms the way plain
    /// HTTP requests do) -- auth now travels the way the real hub actually expects it (an
    /// `access_token` query parameter, the same mechanism browsers use since they can't set
    /// custom headers on a WebSocket upgrade either).
    public func connectWebSocket() async throws {
        guard !isConnected else { return }

        let token = try await negotiate()

        guard var components = URLComponents(
            url: config.websocketURL.appendingPathComponent("/hubs/chat"),
            resolvingAgainstBaseURL: false
        ) else {
            throw ErghiError.webSocketError("Invalid hub URL")
        }
        let accessToken = config.apiKey ?? auth.accessToken ?? ""
        components.queryItems = [
            URLQueryItem(name: "id", value: token),
            URLQueryItem(name: "access_token", value: accessToken),
        ]
        guard let url = components.url else {
            throw ErghiError.webSocketError("Invalid hub URL")
        }

        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            self.handshakeCompleted = false
            self.handshakeContinuation = continuation
            let webSocket = WebSocket(request: URLRequest(url: url))
            webSocket.delegate = self
            self.webSocket = webSocket
            webSocket.connect()
        }

        startKeepalive()
    }

    private struct NegotiateResponse: Decodable {
        let connectionToken: String?
        let connectionId: String?
        let url: String?
    }

    /// POST /hubs/chat/negotiate to obtain the connection token the WebSocket URL needs.
    private func negotiate() async throws -> String {
        let response = try await session.request(
            config.apiURL.appendingPathComponent("/hubs/chat/negotiate"),
            method: .post,
            parameters: ["negotiateVersion": 1],
            encoding: URLEncoding(destination: .queryString)
        )
        .validate()
        .serializingDecodable(NegotiateResponse.self)
        .value

        if response.url != nil {
            // Azure SignalR-style redirect to a different endpoint -- not used by this
            // platform's self-hosted deployment, but fail clearly instead of silently
            // connecting to the wrong place if it ever is.
            throw ErghiError.webSocketError("Hub negotiate returned a redirect target; not supported")
        }
        guard let token = response.connectionToken ?? response.connectionId else {
            throw ErghiError.webSocketError("Hub negotiate response had no connection token")
        }
        return token
    }

    private func startKeepalive() {
        keepaliveTask?.cancel()
        keepaliveTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 15_000_000_000)
                guard let self = self, self.isConnected else { continue }
                self.webSocket?.write(string: "{\"type\":6}" + Self.recordSeparator)
            }
        }
    }

    /// Disconnect from the hub.
    public func disconnectWebSocket() {
        keepaliveTask?.cancel()
        keepaliveTask = nil

        for (_, continuation) in pendingInvocations {
            continuation.resume(throwing: ErghiError.webSocketError("Hub disconnected"))
        }
        pendingInvocations.removeAll()

        webSocket?.disconnect()
        webSocket = nil
        isConnected = false
    }

    /// Invoke a real hub method (e.g. "SendTyping", "JoinConversation", "MarkAsRead").
    ///
    /// When `waitForResult` is true (the default), waits for the server's Completion message
    /// and throws `ErghiError.hubException` if the server rejected the call -- notably
    /// including ConversationOwnershipHubFilter denying access to a conversation outside the
    /// caller's own workspace, which the previous transport could never have surfaced since it
    /// never spoke to a real hub at all.
    @discardableResult
    public func invoke(_ method: String, arguments: [Any], waitForResult: Bool = true) async throws -> Any? {
        guard isConnected, let webSocket = webSocket else {
            throw ErghiError.webSocketError("Hub is not connected")
        }

        var message: [String: Any] = ["type": 1, "target": method, "arguments": arguments]
        var invocationId: String?
        if waitForResult {
            let id = UUID().uuidString
            invocationId = id
            message["invocationId"] = id
        }

        guard let jsonData = try? JSONSerialization.data(withJSONObject: message),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            throw ErghiError.webSocketError("Failed to encode hub invocation")
        }

        guard let invocationId = invocationId else {
            webSocket.write(string: jsonString + Self.recordSeparator)
            return nil
        }

        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any?, Error>) in
            self.pendingInvocations[invocationId] = continuation
            webSocket.write(string: jsonString + Self.recordSeparator)
        }
    }

    /// Send typing indicator. Fire-and-forget, matching this method's previous behavior.
    public func sendTyping(conversationId: String) {
        guard isConnected else { return }
        Task {
            _ = try? await invoke("SendTyping", arguments: [conversationId], waitForResult: false)
        }
    }

    /// Join a conversation's real-time group -- required before UserTyping/MessageRead events
    /// for that conversation are delivered to this connection. Throws
    /// `ErghiError.hubException` if the conversation isn't in the caller's own workspace.
    public func joinConversation(_ conversationId: String) async throws {
        guard isConnected else { return }
        _ = try await invoke("JoinConversation", arguments: [conversationId])
    }

    /// Leave a conversation's real-time group.
    public func leaveConversation(_ conversationId: String) async throws {
        guard isConnected else { return }
        _ = try await invoke("LeaveConversation", arguments: [conversationId], waitForResult: false)
    }

    /// Authenticate a visitor using a signed JWT from the customer's backend.
    public func authenticateVisitor(widgetId: String, jwtToken: String) async throws -> String {
        struct IdentityRequest: Encodable {
            let widgetId: String
            let jwtToken: String
        }

        struct IdentityResponse: Decodable {
            let visitorId: String?
            let VisitorId: String?

            var resolvedId: String? { visitorId ?? VisitorId }
        }

        let request = session.request(
            config.apiURL.appendingPathComponent("/api/conversations/identity"),
            method: .post,
            parameters: IdentityRequest(widgetId: widgetId, jwtToken: jwtToken),
            encoder: JSONParameterEncoder.default
        )

        let response = try await request
            .validate()
            .serializingDecodable(IdentityResponse.self)
            .value

        guard let vId = response.resolvedId else {
            throw ErghiError.authenticationFailed("Visitor ID not found in identity response.")
        }

        self.visitorId = vId
        return vId
    }
}

// MARK: - WebSocketDelegate

extension ErghiClient: WebSocketDelegate {
    public func didReceive(event: Starscream.WebSocketEvent, client: Starscream.WebSocketClient) {
        switch event {
        case .connected:
            if config.debug {
                print("Hub transport connected, sending handshake")
            }
            webSocket?.write(string: "{\"protocol\":\"json\",\"version\":1}" + Self.recordSeparator)

        case .disconnected(let reason, let code):
            isConnected = false
            failPendingInvocations(ErghiError.webSocketError("Hub disconnected: \(reason) (code: \(code))"))
            if config.debug {
                print("Hub disconnected: \(reason) (code: \(code))")
            }

        case .text(let string):
            handleIncomingText(string)

        case .error(let error):
            isConnected = false
            let message = error?.localizedDescription ?? "unknown error"
            failPendingInvocations(ErghiError.webSocketError(message))
            if let handshakeContinuation = handshakeContinuation {
                self.handshakeContinuation = nil
                handshakeContinuation.resume(throwing: ErghiError.webSocketError(message))
            }
            if config.debug {
                print("Hub error: \(message)")
            }

        default:
            break
        }
    }

    private func failPendingInvocations(_ error: Error) {
        for (_, continuation) in pendingInvocations {
            continuation.resume(throwing: error)
        }
        pendingInvocations.removeAll()
    }

    private func handleIncomingText(_ text: String) {
        let records = text.components(separatedBy: Self.recordSeparator).filter { !$0.isEmpty }
        for record in records {
            if !handshakeCompleted {
                handshakeCompleted = true
                handleHandshakeResponse(record)
            } else {
                handleHubRecord(record)
            }
        }
    }

    private func handleHandshakeResponse(_ record: String) {
        guard let data = record.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            resumeHandshake(throwing: ErghiError.webSocketError("Malformed handshake response"))
            return
        }

        if let error = json["error"] as? String {
            resumeHandshake(throwing: ErghiError.webSocketError("Hub handshake failed: \(error)"))
            return
        }

        isConnected = true
        resumeHandshake(throwing: nil)
    }

    private func resumeHandshake(throwing error: Error?) {
        guard let continuation = handshakeContinuation else { return }
        handshakeContinuation = nil
        if let error = error {
            continuation.resume(throwing: error)
        } else {
            continuation.resume()
        }
    }

    private func handleHubRecord(_ record: String) {
        guard let data = record.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? Int else {
            return
        }

        switch type {
        case 1:
            // Invocation from the server -- a hub broadcast (MessageReceived, UserTyping, ...).
            guard let target = json["target"] as? String,
                  let arguments = json["arguments"] as? [Any] else {
                return
            }
            if target == "MessageReceived", let first = arguments.first as? [String: Any] {
                handleMessageReceived(first)
            }

        case 3:
            // Completion -- the response to one of our own invoke() calls.
            guard let invocationId = json["invocationId"] as? String,
                  let continuation = pendingInvocations.removeValue(forKey: invocationId) else {
                return
            }
            if let error = json["error"] as? String {
                continuation.resume(throwing: ErghiError.hubException(error))
            } else {
                continuation.resume(returning: json["result"])
            }

        case 7:
            // Server-initiated close.
            isConnected = false
            if config.debug, let error = json["error"] as? String {
                print("Hub sent close: \(error)")
            }

        default:
            // Type 6 (ping) needs no application-level response beyond our own keepalive task.
            // Types 2/4/5 (streaming) are not used by ChatHub/VisitorChatHub.
            break
        }
    }

    private func handleMessageReceived(_ payload: [String: Any]) {
        guard let messageJson = try? JSONSerialization.data(withJSONObject: payload) else {
            return
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        if let message = try? decoder.decode(Message.self, from: messageJson) {
            messageSubject.send(message)
        }
    }
}
